### ISSUE TITLE
const does not block-scope in loop-body

### ISSUE THREAD

**Issue by oberhamsi** [mozilla/rhino#326]
re-entering the loop-body should create a new const but the old one is re-used:

    Rhino 1.7.7.2 2017 08 24
    js> for (var i=0; i < 8; i++) {
       const x = i * 2;
       print(i, x);
    }  >   >   > 
    0 0
    1 0
    2 0
    3 0
    4 0
    5 0
    6 0
    7 0

expected output (e.g. chrome):

    for (var i=0; i < 8; i++) {
           const x = i * 2;
           console.log(i, x);
        }
    0 0
    1 2
    2 4
    3 6
    4 8
    5 10
    6 12
    7 14

**Comment by gbrail** [comment]
I see the problem. We need to see if anyone here has time to dig in to the parsing and compiling code and see what's happening here.

**Comment by eighthjouster** [comment]
Wow, this is a serious bug.  It's been more than a year since it was reported.  Let's see if I can do something about it.  No promises, though...

**Comment by oberhamsi** [comment]
that would be great! i got used to not using const in that situation

**Comment by jsoref** [comment]
A starting point for someone interested in this would be:
https://github.com/mozilla/rhino/blob/4bd523e36c8ab531841c35ae20408abe51ead004/src/org/mozilla/javascript/Parser.java#L2258-L2260

**Comment by RuralHunter** [comment]
I think this might be related. The code below doesn't compile in rhino:
```javascript
function x () { if (true) { const a = 1 } else { const a = 2 }}
```
rhino reports: net.sourceforge.htmlunit.corejs.javascript.EvaluatorException: TypeError: redeclaration of const a

**Comment by tonygermano** [comment]
From what I can tell, `const` in Rhino is function-scoped like `var` rather than block-scoped like `let`

**Comment by botic** [comment]
I saw a lot of bugs in the last months because of this. If you work with modern JS in a Node-based environment (e.g. frontend stuff), you get used to `const` as block-level variable. Back in Rhino-based code I write `const` like in Node and these error is very hard to spot early.

Any chance to get a better behavior in the next Rhino releases? I fully understand that this is a bigger change, but imho it's getting painful now.

**Comment by tonygermano** [comment]
> Correct, someone would have to start by changing the code I pointed to (including a version flag, because there's a version of JS where that's correct) and then the places where past the parser to actually implement the modern behavior.

@jsoref can you point to the version of JS where that was the correct behavior? I looked through the past specs, and couldn't find evidence of when it was not block scoped. Maybe it was some draft version?

**Comment by botic** [comment]
`const` became a reserved word in JavaScript 1.5 and afais it has been part of SpiderMonkey with this release.

**Comment by p-bakker** [comment]
Since the semantics of `const` are near identical to the semantics of `let`, the main differentiator being is that a `const` cannot be reassigned, I feel fixing the `const` implementation shouldn't be that hard anymore, assuming the `let` implementation is (made) standards compliant.

There's at least https://github.com/mozilla/rhino/issues/923 to make the `let` implementation more standards compliant

**Comment by p-bakker** [comment]
To my knowledge, the current behavior of `const` was never in a finalized spec: maybe the behavior was in a draft spec, but never like this in a final version of EcmaScript

**Comment by tuchida** [comment]
Rhino `let`/`const` does not implement Temporal dead zone (TDZ). If we try to fully comply with the spec, we expect to have to modify it quite a bit.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz

**Comment by p-bakker** [comment]
Another (potential?) non-compliant issue in the `let` implementation: https://github.com/mozilla/rhino/issues/834

**Comment by rbri** [comment]
We have a bunch of const related issue and i can remember that i have tried to fix some. But  in the end the backward compatibility was always the blocker. I think the suggestion for starting a branch that implements the current spec without taking care of backward compatibility was based on this experiences.

**Comment by FlyingYu-Z** [comment]
Serious bug still unresolved untill 2024.

**Comment by qtlunya** [comment]
> We have a bunch of const related issue and i can remember that i have tried to fix some. But in the end the backward compatibility was always the blocker. I think the suggestion for starting a branch that implements the current spec without taking care of backward compatibility was based on this experiences.

Why is backward compatibility with behavior that was never part of the spec more important than complying with the current spec that has existed for nearly a decade now?

**Comment by p-bakker** [comment]
> Why is backward compatibility with behavior that was never part of the spec more important than complying with the current spec that has existed for nearly a decade now?

Backwards compatibility isn't an issue (anymore) as there are several ways to deal with that. 

The main reason this hasn't been resolved yet is time I guess., time to dig in and create a MR that implemented the spec-compliant behaviour. 

Rhino is maintained by volunteers (mostly), no major organization that actively uses Rhino (of which there are quite a few) is contributing dedicated resources to the project, so progress is made when PRs are provided.

**Comment by qtlunya** [comment]
Yeah, fair enough. I'm not the one using Rhino directly. I'm using [ChatTriggers](https://github.com/chattriggers), which uses Rhino as its JavaScript interpreter to effectively allow scripting Minecraft with JavaScript. I'm not sure how feasible it would be for them to switch to a different interpreter, but this bug with `const` comes up relatively frequently in their server when plugin developers run into it.

**Comment by gbrail** [comment]
Thanks for the discussion here, and I’m excited to know that Rhino is used
in Minecraft!

Yeah, the “const” stuff is one of those things where I think it’s going to
be quite a bit of work to fix, and AFAIK there is not a person in the world
who is actually paid to maintain Rhino. (although “rbri” uses it in his
flagship project and is a very frequent contributor!)

So anyone who thinks that they can start to tackle fixing constness is
quite welcome, and I’m happy to discuss backward compatibility policies if
it helps make that move forward faster.

**Comment by p-bakker** [comment]
I started on it quite a while ago, its relatively doable (at least in interpreted mode, haven't looked at compiled mode), but lacking time to continue to work on it for now. 

As const are very much like let (for which Rhino has pretty decent support) with regards to scoping rules. Main difference is that const can only be assigned on declaration

**Comment by tonygermano** [comment]
I'd be happy to ditch backwards compatibility on this one, too. I can think of two situations where backward compatibility is relevant.

1. This code should execute in a spec compliant runtime
```js
const myVar = 1
{
    const myVar = 2
    assert(myVar === 2)
}
assert(myVar === 1)
```
Rhino will throw `js: uncaught JavaScript runtime exception: TypeError: TypeError: redeclaration of const myVar.`

There cannot be any code running successfully with this TypeError, and I don't see a reason to preserve it.

2. This code will throw a `Uncaught TypeError: Assignment to constant variable.` in a spec compliant runtime.
```js
const myVar = 1
{
    myVar = 2
    assert(myVar === 1)
}
assert(myVar === 1)
```
Rhino will silently not do the assignment and keep running. This seems dangerous and unexpected to me and a likely source of bugs. I think it's highly unlikely that anyone would be depending on this behavior.

**Comment by SteveHawes** [comment]
Sad to see there is a major new release and still no movement on this issue

**Comment by jcompagner** [comment]
i added in our fork a patch for this, including a feature flag to set back the old behavior:

https://github.com/Servoy/rhino/commit/e5387d851e8e5c68b3595408961f01fd01d16b04

I can make a pull request for this if needed on the current master of rhino

If anybody has some comments please let me know

I didn't test the compiled mode, but the changes are not really in the Intepreter (only really in the parser/transformer and the ScriptRuntime)

**Comment by aardvark179** [comment]
We have done quite a bit of discussion on this internally, and have a fairly solid plan for resolving this. With our current activation frame we pay a significant performance penalty compared to simple local variable access. Worse, because scopes can be captured by closures, you need a new nested frame on every loop iteration.

To resolve this without significant performance issues we first need to improve scope and variable handling in the parser so we can number variables in scopes and refer to them by how far up in that hierarchy we need to go along with the number, then we can better analyse when activation frames are actually required and we will have what we need to make them fast.

**Comment by jcompagner** [comment]
i don't see where the real performance issue is here
You talk about that we need nested frames, but that is not the way how it works
Because what we create now is NativeWith with an ObjectLiteral creation that is the scope object of that.. (that is what i fixed above that the ObjectListeral creation understand that it needs to defineConst() objects. not just put of a variable/let

Currently we don't create frames because then the (current) debugger would work, but the debugger doesn't work for nested blocks that define lets (and now const) for that i updated the interface:
https://github.com/Servoy/rhino/commit/a47050a9f7def9a4bdb751d8a1cba90d36a92371#diff-8bb6c202383af88ba8588ada508e5749042bb1d6a79f922858d4405094ce33e6

so that when a NativeWith is created we let the debugger know this so that it knows what now the top scope stack is.

So our implementation of the Debugger interface now knows what really the current scope object is and we see nicely the lets and the const popping in and out when entering or leaving a scoped block.

So i wonder where in your eyes the performance penalty really is, because as far as i see its kind of just the creation of a NativeWith object with a NativeObject as its scope.. (and i don't think in any kind of implementation you will get around that)

**Comment by aardvark179** [comment]
Okay, there's a bit to unpack here. Let's start with local variable usage vs. frames.

Whether it's a `NativeCall` or `NativeWith` that's being created isn't really important. The important thing is the shift from our raw local variab.le storage to using _anything_ that requires access by name rather than by simple offset. Let me illustrate with a simple benchmark
```javascript
function doWork2(iterations) {
    var x = iterations;
    var y = 0.5;
    for (var i = 0; i < x; i++) {
        y++;
    }
    return y;
}
```
Then for 10,000 iterations this gives the following on my machine
```
Benchmark           (interpreted)  Mode  Cnt    Score    Error  Units
TestBenchmark.test          false  avgt    5   10.969 ±  0.029  us/op
TestBenchmark.test           true  avgt    5  394.495 ± 22.919  us/op
```
Now, if I change that to require an activation frame and hence forcing variables to be referenced by name (in this case by referencing the `arguments` object) then we see the following:

```
Benchmark           (interpreted)  Mode  Cnt    Score    Error  Units
TestBenchmark.test          false  avgt    5  219.112 ± 23.988  us/op
TestBenchmark.test           true  avgt    5  562.451 ±  4.644  us/op
```

That's a pretty big slowdown! So why does it happen?

There's two things contributing to this. First is that simple local variable lookup is a simple indexed lookup, while lookup in any sort of variable frame currently requires a hash table lookup every time, and the other issue is that local variables can be stored as either objects or raw double values, with the latter avoiding a boxing an operation, while frames require boxing on every store. You can see these operations in profiles. So, we need to be careful about increasing the amount of code that requires frames to be created as it _will_ cause a performance regression.

The other issue, and why you may need a separate frame on every iteration of a loop is how variables can be captured. If we have a simple function like this:

```javascript
function f(iterations) {
    var a = [];
    var x = 0;
    for (x = 0; x < iterations; x++) {
        a[x] = () => x;
    }
    return a;
}
```
Then `f(5)` should return an array of five functions, and each one of those functions will return `x` from the captured frame. That value will be `5` because we only have one frame, and by the time we exit the function `x` will have been incremented to `5`. So `f(5)[3]()` will return `5`, and so will `f(5)[4]()`, etc.  If we change this to be

```javascript
function f(iterations) {
    var a = [];
    var x = 0;
    for (x = 0; x < iterations; x++) {
        const y = x;
        a[x] = () => y;
    }
    return a;
}
```

Then `f(5)[1]()` will return `1`, `f(5)[2]()` will return 2, etc. Each iteration of the loop creates a new declaration environment which might be captured by a closure.

These two aspects are why I would suggest that the right order in which to tackle this is to first improve local variable handling in the parser to avoid the need to use frames, and to also improve how those frames work so they can also be a simple indexed lookup rather than a hash lookup. If we track scopes, and which might be captured by `eval` or closures in the parser / IR, then we can implement `const` and `let` correctly without creating frames (instead we could mark the variables as UNINITIALISED, and OUT_OF_SCOPE as appropriate), and we can mostly avoid creating frames entirely.

To address your last question about how we can avoid the costs of  `NativeWith` and `NativeObject`, or something similar let me expand on how variables and scopes are handled. Objects in Javascript are mutable in a way that scopes are not, at least in general. I can might start with an empty object `{ }` and later add properties over time, but in general I know the complete list of variables in a declaration scope at the time of compilation. When we're simply dealing with local variables and don't create an activation frame then we rely on this in the compiler—we number each local variable and access them by this index. If we extend this to activation frames and other nested declaration scopes then we can avoid almost all of the overhead associated with those without degrading performance but to do that we do need somewhere convenient to store information about the shape of those frames (the recent work to refactor function objects was partially motivated by that, the shapes of the declaration scopes can be easily stored as part of the declaration objects). The exception to this would be `with` scopes which make a scope out of mutable object. We use those in a number of places in Rhino at the moment, but this is more an accident of history rather than a real requirement, and they are something we should try and move away from except for the compilation of `with` statements.

**Comment by jcompagner** [comment]
ah thx for the info.

I guess for now we keep the implementation then that we have, and see later. 
Because we where getting constant support cases asking this question why lets and const behave like they do , and then also different based on if you have a nested function or not... (and the debugger behaves differently also because of this)

This case is already 8 years old, so i rather have a fix now that kind of works for the most people the same (a nested function is done very quickly so then the activation is already set to true any way for most of our users..)

Do check if you change all this also the debugger. Because that now works with Frames and that is called as far as i see only for when the function is called., but you have to go into nested scopes somehow and let also the debugger know that that suddenly there are more lets/const (or less when you exit that scope)

**Comment by aardvark179** [comment]
Yes. The debugger definitely needs work. It currently isn't even tests properly either for execution behaviour (i.e. do all tests pass with the debugger enabled), and or for functional behaviour (are the correct calls made to the debugger when various events happen). We need to improve both of those and I hope we can create some PRs in that area soon.

Slightly longer term we really need to rethink the debugger interface as the current version is a poor match to modern debugging needs. Its line based design does not fit well with bundling, minification, and source maps, and I haven't had time to play with it in combination with things like generator functions to see if it does the right thing.

**Comment by aardvark179** [comment]
I absolutely understand your frustration, and that of your customers. We took a different approach and used transpilation, which can turn the sort of const loop above into
```javascript
function f(iterations) {
    function _loop() {
        var p = x;
        a[x] = () => p;
    }
    var a = [];
    var x = 0;
    for (x = 0; x < iterations; x++) {_loop()}
    return a;
}
```
We still pay all the performance price of having activation frames, but it does give us correct behaviour while we work to improve Rhino itself, and with some careful use of source maps still allows for the use of the debugger without showing the user any transpired code.
